src/main/java/com/amihaiemil/camel/WellIndentedLine.java

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (c) 2016-2017, Mihai Emil Andronache
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ *  list of conditions and the following disclaimer.
+ *  Redistributions in binary form must reproduce the above copyright notice,
+ *  this list of conditions and the following disclaimer in the documentation
+ *  and/or other materials provided with the distribution.
+ * Neither the name of the copyright holder nor the names of its
+ *  contributors may be used to endorse or promote products derived from
+ *  this software without specific prior written permission.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+package com.amihaiemil.camel;
+
+/**
+ * Decorator for YamlLine.
+ * @author Sherif Waly (sherifwaly95@gmail.com)
+ * @version $Id$
+ * @since 1.0.0
+ *
+ */
+public final class WellIndentedLine implements YamlLine {
+
+    /**
+     * Content line.
+     */
+    private YamlLine line;
+    
+    /**
+     * Previous line.
+     */
+    private YamlLine previousLine;
+    
+    /**
+     * Ctor.
+     * @param line YamlLine
+     * @param previousLine YamlLine
+     */
+    public WellIndentedLine(final YamlLine line, final YamlLine previousLine) {
+        this.line = line;
+        this.previousLine = previousLine;
+    }
+    
+    @Override
+    public String trimmed() {
+        return this.line.trimmed();
+    }
+    
+    @Override
+    public int number() {
+        return this.line.number();
+    }
+
+    @Override
+    public int indentation() {
```

Comment:
You got the idea, but this method should be a little more complex, to cover the several cases of indentation.

What I mean is that, in YAML, indentation matters. For instance, if the current line ends with ``:``, then the next one should have a +2 indentation. e.g.

```
mappingArray:
  - elem1
  - elem2
```
On the other hand, if the line doesn't end in a special character, then the next line should have the same or lower indentation (e.g. ``- elem1`` and ``- elem2`` have the same indentation).

The following chars indicate that the next line should have a +2 indentation: ``: > | -``. 
You could add them in a list and do the following logic (pseudocode):

```
List specialChars = [':', '>', '|', '-'];
if (specialChars.contains(this.previousLine.trimmed().lastChar()) {
    if (lineIndent - previousLineIndent != 2) {
        throw Exceptionl
    }
} else {
    if (lineIndent > previousLineIndent) {
        throw Excption
    }
}
return lineIndent;
```

Makes sense? :D 
And add a puzzle for unit tests, I think you will have already spend more than 30 min.